source/core/lethegridtools.cc

```diff
@@ -0,0 +1,905 @@
+      // stop and loop if the cell is active if the cell is not active loop
+      // on the child of the cell. Repeat
+      unsigned int lvl        = 0;
+      unsigned int max_childs = GeometryInfo<dim>::max_children_per_cell;
+      while (best_cell_iter->is_active() == false)
+        {
+          bool         cell_found = false;
+          double       best_dist  = DBL_MAX;
+          unsigned int best_index = 0;
+          for (unsigned int i = 0; i < max_childs; ++i)
+            {
+              try
+                {
+                  const Point<dim, double> p_cell =
+                    mapping.transform_real_to_unit_cell(
+                      best_cell_iter->child(i), point);
+                  const double dist =
+                    GeometryInfo<dim>::distance_to_unit_cell(p_cell);
+                  bool inside = true;
+
+
+                  if (dist <= best_dist and inside)
+                    {
+                      best_dist  = dist;
+                      best_index = i;
+                      cell_found = true;
+                      if (dist == 0)
+                        break;
+                    }
+                }
+              catch (
+                const typename MappingQGeneric<dim>::ExcTransformationFailed &)
+                {}
+            }
+
+          best_cell_iter = best_cell_iter->child(best_index);
+
+          if (cell_found == false)
+            {
+              std::cout << "Cell not found around " << point << std::endl;
+              break;
+            }
+
+          lvl += 1;
+          best_dist_last = best_dist;
+        }
+    }
+  if (best_dist_last >= 1e-9)
+    {
+      throw "The point is not inside the mesh";
+    }
+
+
+  return best_cell_iter;
+}
+
+template <int dim>
+typename DoFHandler<dim>::active_cell_iterator
+LetheGridTools::find_cell_around_point_with_neighbors(
+  const DoFHandler<dim> &dof_handler,
+  std::map<unsigned int,
+           std::set<typename DoFHandler<dim>::active_cell_iterator>>
+                                                       &vertices_cell_map,
+  const typename DoFHandler<dim>::active_cell_iterator &cell,
+  const Point<dim>                                     &point)
+{
+  // Find the cells around the initial cell ( cells that share a vertex with the
+  // original cell).
+  std::vector<typename DoFHandler<dim>::active_cell_iterator>
+    active_neighbors_set =
+      LetheGridTools::find_cells_around_cell<dim>(vertices_cell_map, cell);
+  // Loop over that group of cells
+  for (unsigned int i = 0; i < active_neighbors_set.size(); ++i)
+    {
+      bool inside_cell = active_neighbors_set[i]->point_inside(point);
+      if (inside_cell)
+        {
+          return active_neighbors_set[i];
+        }
+    }
+  // The cell is not found near the initial cell so we use the cell tree
+  // algorithm instead (much slower).
+  std::cout << "Cell not found around " << point << std::endl;
```

Comment:
Try to avoid using std::cout in the code, what happens if it running in parallel? It will print like crazy.

Reply:
This will append only in one process as it's related to only one point. So when there is a problem only one proc shows this message.